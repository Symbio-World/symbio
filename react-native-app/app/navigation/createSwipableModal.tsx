import * as React from 'react'
import { StackNavigationProp } from '@react-navigation/stack'
import { Dimensions, View, StyleSheet } from 'react-native'
import { t } from 'react-native-tailwindcss'
import Animated from 'react-native-reanimated'

import { Interactable } from '../navigation/Interactable'
import { ModalCard } from './ModalCard'

import { ModalStackParamList } from './Navigation'

const { Value, interpolate, concat, Extrapolate } = Animated
const { width, height } = Dimensions.get('window')
const deltaX = width / 2
const α = Math.PI / 12
const A = width * Math.cos(α) + height * Math.sin(α)

type SwipableModalProps = {
  onSnap: (e: { nativeEvent: { x: number } }) => void
}

// test require this component for identifing by type
export const SwipableModal: React.FC<SwipableModalProps> = ({
  children,
  onSnap,
}) => {
  const x = new Value(0)
  const y = new Value(0)
  const rotateZ = concat(
    interpolate(x, {
      inputRange: [-1 * deltaX, deltaX],
      outputRange: [α, -1 * α],
      extrapolate: Extrapolate.CLAMP,
    }),
    'rad',
  )

  const translateX = x
  const translateY = y

  const style = [
    StyleSheet.absoluteFill,
    {
      transform: [{ translateX }, { translateY }, { rotateZ }],
    },
  ]

  return (
    <View style={[t.flex1]}>
      <Interactable
        style={style}
        snapPoints={[{ x: -1 * A }, { x: 0 }, { x: A }]}
        onSnap={onSnap}
        x={x}
        y={y}
      >
        <ModalCard>{children}</ModalCard>
      </Interactable>
    </View>
  )
}
type NavigationProps = {
  navigation: StackNavigationProp<ModalStackParamList>
}

type CreateSwipableModal = <Props extends NavigationProps>(
  WrapperCompponent: React.FC<Props>,
) => React.FC<Props>

export const createSwipableModal: CreateSwipableModal = (Component) => {
  return (props) => {
    const handleSnap = ({
      nativeEvent: { x },
    }: {
      nativeEvent: { x: number }
    }) => {
      const isDefaulPosition = x === 0
      if (!isDefaulPosition) {
        props.navigation.goBack()
      }
    }

    return (
      <SwipableModal onSnap={handleSnap}>
        <View style={[t.hFull]}>
          <Component {...props} />
        </View>
      </SwipableModal>
    )
  }
}
