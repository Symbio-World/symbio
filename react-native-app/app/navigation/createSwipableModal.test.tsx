import * as React from 'react'
import { render, fireEvent } from 'react-native-testing-library'
import { View, Text } from 'react-native'

import { createSwipableModal, SwipableModal } from './createSwipableModal'

jest.mock('react-native-reanimated', () =>
  require('react-native-reanimated/mock'),
)

describe('createSwipableModal', () => {
  const navigation: any = {
    goBack: jest.fn(),
  }
  const SwipableModalComp = createSwipableModal(() => (
    <View>
      <Text>Test</Text>
    </View>
  ))
  it('renders correctly', () => {
    const { toJSON } = render(<SwipableModalComp navigation={navigation} />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('do nothing on default position', () => {
    const { getByType } = render(<SwipableModalComp navigation={navigation} />)
    const eventData = {
      nativeEvent: {
        x: 0,
      },
    }
    fireEvent(getByType(SwipableModal), 'onSnap', eventData)
    expect(navigation.goBack).toBeCalledTimes(0)
  })

  it('call go back on snap', () => {
    const { getByType } = render(<SwipableModalComp navigation={navigation} />)
    const eventData = {
      nativeEvent: {
        x: 2,
      },
    }
    fireEvent(getByType(SwipableModal), 'onSnap', eventData)
    expect(navigation.goBack).toBeCalled()
  })
})
