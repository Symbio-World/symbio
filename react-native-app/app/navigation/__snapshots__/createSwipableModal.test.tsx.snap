// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSwipableModal renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
    ]
  }
>
  <View
    collapsable={false}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "transform": Array [
            Object {
              "translateX": Object {
                "setValue": [Function],
              },
            },
            Object {
              "translateY": Object {
                "setValue": [Function],
              },
            },
            Object {
              "rotateZ": "undefinedrad",
            },
          ],
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0,0,0,0)",
          },
          Object {
            "width": "100%",
          },
          Object {
            "height": "100%",
          },
          Object {
            "alignItems": "center",
          },
          Object {
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "width": "91.666667%",
            },
            Object {
              "height": "95%",
            },
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "borderRadius": 8,
            },
            Object {
              "elevation": 12.5,
              "shadowColor": "rgba(0, 0, 0, 0.1)",
              "shadowOffset": Object {
                "height": 20,
                "width": 0,
              },
              "shadowOpacity": 1,
              "shadowRadius": 25,
            },
            Object {
              "overflow": "hidden",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "height": "100%",
              },
            ]
          }
        >
          <View>
            <Text>
              Test
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
